whoc/design_tools/wake_steering_design.py

```diff
@@ -0,0 +1,630 @@
+    wd_max: float = 360,
+    ws_resolution: float = 1,
+    ws_min: float = 8,
+    ws_max: float = 8,
+    ti_resolution: float = 0.02,
+    ti_min: float = 0.06,
+    ti_max: float = 0.06,
+    minimum_yaw_angle: float = 0.0,
+    maximum_yaw_angle: float = 25.0,
+    kwargs_UncertainFlorisModel: dict = {},
+) -> pd.DataFrame:
+    """
+    Build a simple wake steering lookup table for a given FlorisModel using the Serial Refine
+    method, with uncertainty in the wind direction.
+
+    This is mainly intended for demonstration purposes. Users that want more control over the
+    optimization process should use FLORIS's YawOptimizationSR procedure directly to produce
+    the output df_opt.
+
+    Args:
+        fmodel (FlorisModel): An instantiated FlorisModel object.
+        wd_std (float): Wind direction standard deviation in degrees.
+        wd_resolution (float, optional): The resolution of the wind direction in degrees.
+            Defaults to 5.
+        wd_min (float, optional): The minimum wind direction in degrees. Defaults to 0.
+        wd_max (float, optional): The maximum wind direction in degrees. Defaults to 360.
+        ws_resolution (float, optional): The resolution of the wind speed in m/s.
+            Defaults to 1.
+        ws_min (float, optional): The minimum wind speed in m/s. Defaults to 8.
+        ws_max (float, optional): The maximum wind speed in m/s. Defaults to 8.
+        ti_resolution (float, optional): The resolution of the turbulence intensity as a fraction.
+            Defaults to 0.02.
+        ti_min (float, optional): The minimum turbulence intensity as a fraction. Defaults to 0.06.
+        ti_max (float, optional): The maximum turbulence intensity as a fraction. Defaults to 0.06.
+        minimum_yaw_angle (float, optional): The minimum allowable misalignment in degrees.
+            Defaults to 0.0.
+        maximum_yaw_angle (float, optional): The maximum allowable misalignment in degrees.
+            Defaults to 25.0.
+        kwargs_UncertainFlorisModel (dict, optional): Additional keyword arguments for the
+            instantiation of the UncertainFlorisModel. Defaults to an empty dictionary.
+
+    Returns:
+        pd.DataFrame: A yaw offset lookup table.
+    """
+    wind_rose = create_uniform_wind_rose(
+        wd_resolution=wd_resolution,
+        wd_min=wd_min,
+        wd_max=wd_max,
+        ws_resolution=ws_resolution,
+        ws_min=ws_min,
+        ws_max=ws_max,
+        ti_resolution=ti_resolution,
+        ti_min=ti_min,
+        ti_max=ti_max,
+    )
+
+    fmodel.set(wind_data=wind_rose)
+
+    ufmodel = UncertainFlorisModel(
+        configuration=fmodel.core.as_dict(),
+        wd_std=wd_std,
+        **kwargs_UncertainFlorisModel,
+    )
+
+    yaw_opt = YawOptimizationSR(
+        fmodel=ufmodel,
+        minimum_yaw_angle=minimum_yaw_angle,
+        maximum_yaw_angle=maximum_yaw_angle,
+    )
+
+    return yaw_opt.optimize()
+
+
+def apply_static_rate_limits(
```

Comment:
Made the comment elsewhere but wanted to find a way this method is probably not preferred to putting limits in the time domain.  This is forcing the controller to sub-optimal angles even if the wind direction changes slowly (freely admit this was I think my idea originally...)